state-synchronizer/src/request_manager.rs

```diff
@@ -234,12 +233,33 @@ impl RequestManager {
         None
     }
 
-    pub fn pick_peers(&self) -> Vec<PeerNetworkId> {
-        self.eligible_peers
-            .iter()
-            .take(self.multicast_level)
-            .filter_map(|(_, (peers, weighted_index))| Self::pick_peer(peers, weighted_index))
-            .collect::<Vec<_>>()
+    pub fn pick_peers(&mut self) -> Vec<PeerNetworkId> {
+        // Strategy: pick peers using multicast level
+        // if no live peers exist for this multicast level, keep failing over to next level
+
+        let mut chosen_peers = vec![];
+        for (level, (peers, weighted_index)) in self.eligible_peers.iter() {
+            if let Some(peer) = Self::pick_peer(peers, weighted_index) {
+                chosen_peers.push(peer)
+            }
+            // at the minimum go through networks with preference level <= multicast_level
+            // if no peers are found for the current multicast_level, continue doing
+            // best effort search of the first network with live peers to failover to
+            if !chosen_peers.is_empty() && *level >= self.multicast_level {
+                if *level > self.multicast_level {
+                    info!(
+                        LogSchema::event_log(LogEntry::Multicast, LogEvent::Failover)
+                            .old_multicast_level(self.multicast_level)
+                            .new_multicast_level(*level),
+                    );
+                    self.multicast_level = *level;
+                    counters::MULTICAST_LEVEL.set(self.multicast_level as i64);
+                }
```

Comment:
I think this logic is a little nuanced and not well documented...

Effectively what you want is one function to make your preference less strict (more open) and that is dependent on lack of availability of a node.

Then you want another function to increase that preference level based upon availability of actual data.

This does have shortcomings, which is namely these two properties are not symmetric and thus hard to cleanly reason about. At any rate, a doc block on this would be great and then you can reference that doc block in the code.

"This relaxes our multicast requirements so that we can incorporate additional peers for making state sync requests."

Also larger level is less preferred